plugins/modules/ipacert.py

```diff
@@ -0,0 +1,382 @@
+
+    return changed, {}
+
+
+def main():
+    ansible_module = IPAAnsibleModule(
+        argument_spec=dict(
+            # requested
+            csr=dict(type="str"),
+            principal=dict(type="str"),
+            add_principal=dict(type="bool"),
+            profile=dict(type="str"),
+
+            # revoked
+            reason=dict(type="str"),
+
+            # general
+            serial_number=dict(type="int"),
+            ca=dict(type="str"),
+
+            # state
+            state=dict(type="str", required=True,
+                       choices=["requested", "held", "released", "revoked"]),
+        ),
+        required_if=[
+            ('state', 'requested', ['csr', 'principal']),
+            ('state', 'held', ['serial_number']),
+            ('state', 'relased', ['serial_number']),
+            ('state', 'revoked', ['serial_number', 'reason']),
+        ],
+        supports_check_mode=False,
+    )
+
+    ansible_module._ansible_debug = True
+
+    # Get parameters
+
+    # requested
+    csr = ansible_module.params_get("csr")
+    principal = ansible_module.params_get("principal")
+    add_principal = ansible_module.params_get("add_principal")
+
+    # revoked
+    reason = ansible_module.params_get("reason")
+
+    # general
+    ca = ansible_module.params_get("ca") # XXX no test coverage
+    serial_number = ansible_module.params.get("serial_number")
+    profile = ansible_module.params_get("profile")
+
+    # state
+    state = ansible_module.params_get("state")
+
+    # Check parameters
+
+    invalid = []
+
+    if state == "requested":
+        if serial_number:
+            invalid.append("serial_number")
+            ansible_module.fail_json(msg="serial_number not used with state=requested")
+        elif reason:
+            invalid.append("reason")
+            ansible_module.fail_json(msg="reason not used with state=requested")
+    elif state == "held" or state == "revoked" or state == "released":
+        if csr:
+            invalid.append("csr")
+            ansible_module.fail_json(msg="csr not used with state=requested")
+        elif principal:
+            invalid.append("principal")
+            ansible_module.fail_json(msg="principal not used with state=requested")
+        elif add_principal:
+            invalid.append("add_principal")
+            ansible_module.fail_json(msg="principal not used with state=requested")
+
+        if state == "held" or state == "released":
+            if reason:
+                invalid.append("reason")
+                ansible_module.fail_json(msg="reason not used with state=held/released")
+            else:
+                reason_int = REVOCATION_REASONS["certificateHold"]
+        else:
+            if reason in REVOCATION_REASONS:
+                reason_int = REVOCATION_REASONS[reason]
+            else:
+                invalid.append("reason")
+                ansible_module.fail_json(msg="reason %s not valid" % (reason,))
```

Comment:
Almost all of the 'ansible_module.fail_json()` here are unnecessary, as the invalid arguments are evaluated later (with `ansible_module.params_fail_used_invalid`).

I suggest something like:
```python
if state == "requested":
    invalid = ["serial_number", "reason"]
elif state == "released":
    invalid = ["csr", "principal", "add_principal"]
    if reason not in REVOCATION_REASONS:
        ansible_module.fail_json(msg="reason %s not valid" % reason)
elif ...
```

Reply:
Looks like we don't even need to the `reason` option explicitly, `ansible_module` does it for us